sdk/go/src/sawtooth_sdk/client/encoder.go

```diff
@@ -0,0 +1,270 @@
+package client
+
+import (
+	"encoding/hex"
+	"fmt"
+	"github.com/golang/protobuf/proto"
+	"sawtooth_sdk/protobuf/batch_pb2"
+	"sawtooth_sdk/protobuf/transaction_pb2"
+	"strings"
+	"time"
+)
+
+type TransactionParams struct {
+	FamilyName      string
+	FamilyVersion   string
+	PayloadEncoding string
+	Nonce           string
+	BatcherPubkey   string
+	Dependencies    []string
+	Inputs          []string
+	Outputs         []string
+}
+
+type Encoder struct {
+	privkey  []byte
+	pubkey   string
+	defaults TransactionParams
+}
+
+// NewTransactionEncoder constructs a new encoder which can be used to generate
+// transactions and batches, and to serialize batches for submitting to the
+// REST API.
+func NewEncoder(privkey []byte, defaults TransactionParams) *Encoder {
+	return &Encoder{
+		privkey:  privkey,
+		pubkey:   Encode(GenPubKey(privkey), "hex"),
+		defaults: defaults,
+	}
+}
+
+// -- Transactions --
+
+// NewTransaction Creates a new transaction and handles the construction and
+// signing of the transaction header.
+func (self *Encoder) NewTransaction(payload []byte, p TransactionParams) *Transaction {
+	h := &transaction_pb2.TransactionHeader{
+		// Load defaults
+		FamilyName:      self.defaults.FamilyName,
+		FamilyVersion:   self.defaults.FamilyVersion,
+		PayloadEncoding: self.defaults.PayloadEncoding,
+		Nonce:           self.defaults.Nonce,
+		BatcherPubkey:   self.defaults.BatcherPubkey,
+
+		Inputs:       self.defaults.Inputs,
+		Outputs:      self.defaults.Outputs,
+		Dependencies: self.defaults.Dependencies,
+
+		// Set unique fields
+		PayloadSha512: Encode(SHA512(payload), "hex"),
+		SignerPubkey:  self.pubkey,
+	}
+
+	// Override defaults if set
+	if p.FamilyName != "" {
+		h.FamilyName = p.FamilyName
+	}
+	if p.FamilyVersion != "" {
+		h.FamilyVersion = p.FamilyVersion
+	}
+	if p.PayloadEncoding != "" {
+		h.PayloadEncoding = p.PayloadEncoding
+	}
+	if p.Nonce != "" {
+		h.Nonce = p.Nonce
+	}
+	if p.BatcherPubkey != "" {
+		h.BatcherPubkey = p.BatcherPubkey
+	}
```

Comment:
Since the field types are different, I had to use a struct instead of a map, so there wasn't a good way I know of. I think of Go as just a prettier version of C with a GC